Cython/Utility/ImportExport.c

```diff
@@ -771,11 +768,32 @@ static int __Pyx_MergeVtables(PyTypeObject *type) {
     // resolution isn't possible and we must reject it just as when the
     // instance struct is so extended.  (It would be good to also do this
     // check when a multiple-base class is created in pure Python as well.)
-    for (i = 1; i < PyTuple_GET_SIZE(bases); i++) {
-        void* base_vtable = __Pyx_GetVtable(((PyTypeObject*)PyTuple_GET_ITEM(bases, i)));
+#if CYTHON_ASSUME_SAME_MACROS
+    size = PyTuple_GET_SIZE(bases);
+#else
+    size = PyTuple_Size(bases);
+    if (size < 0) goto other_failure;
+#endif
+    for (i = 1; i < size; i++) {
+        PyObject *basei;
+        void* base_vtable;
+#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
+        basei = PyTuple_GET_ITEM(bases, i);
+#else
+    #if !CYTHON_AVOID_BORROWED_REFS
+        basei = PyTuple_GetItem(bases, i);
+    #else
+        basei = PySequence_Item(bases, i);
+    #endif
+        if (unlikely(!basei)) goto other_failure;
+#endif
```

Comment:
This can probably be restructured into a simpler sequence of conditions, rather than nesting them. Especially given the separate decref below, which seems non-trivial to reason about.